### PR TITLE
(feature) Add drag and drop functionality to the sprint board

### LINKED ISSUE #42: Make tasks on the sprint board draggable; update their status and rank when dropping
Mockup of sprintboard view
### Sprintboard

turtle_sprint
### Editing a sprint

turtle_sprint_settings
### Creating a new task:

turtle_task

### PR DESCRIPTION
**Pull request by azangru**
closes #42 
closes #63

### CODE REVIEW

File: frontend/src/components/sprintboard/board.jsx [hunk 1]
```diff
@@ -1,46 +1,51 @@
...
     };
   },
 
-  // onStoreUpdate(addressee) {
-  //   this.setState({
-  //     addressee: addressee
-  //   });
-  // },
+  componentDidMount() {
+    this.listenTo(SprintStore, this.onStoreUpdate);
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/board.jsx:25]: I did a little bit of research but I don't think I found an answer - is it common practice to name the listener simply `onStoreUpdate`? If so, I'd like to change mine to the same thing.

File: frontend/src/components/sprintboard/board.jsx [hunk 2]
```diff
@@ -1,46 +1,51 @@
...
+import _ from 'lodash';
 import React from 'react/addons';
-// import Reflux from 'reflux';
-// import Actions from '../../actions/actions';
+import Reflux from 'reflux';
+import {SprintActions} from '../../actions/actions';
+import SprintStore from '../../stores/sprintStore';
 
-import SprintColumn from './column.jsx';
+import { DragDropContext } from 'react-dnd';
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/board.jsx]: let's be consistent - no spaces inside brackets

File: frontend/src/components/sprintboard/column.jsx [hunk 3]
```diff
@@ -1,23 +1,46 @@
 import React from 'react/addons';
+import {SprintActions} from '../../actions/actions';
 import Task from './task.jsx';
 
+import { ItemTypes } from './constants';
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/column.jsx]: `{ItemTypes}`

File: frontend/src/components/sprintboard/column.jsx [hunk 4]
```diff
@@ -1,23 +1,46 @@
...
+import { DropTarget } from 'react-dnd';
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/column.jsx]: `{DropTarget}`

File: frontend/src/components/sprintboard/task.jsx [hunk 5]
```diff
@@ -1,18 +1,78 @@
-import React from 'react/addons';
+import React, { PropTypes } from 'react/addons';
+import {SprintActions} from '../../actions/actions';
+import { ItemTypes } from './constants';
+import { DragSource, DropTarget } from 'react-dnd';
+
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/task.jsx:5]: no spaces in brackets in lines 1-4

File: frontend/src/components/sprintboard/board.jsx (hunk 1 above)
Comment (azangru) [inline frontend/src/components/sprintboard/board.jsx:25]: I don't know whether it is common practice nor do I remember where I borrowed this name, but it feels suitable. If I am not mistaken, the relationship between a component and a store is pretty dumb: store fires only one kind of event, and component listens to it. So basically there is only one function in the component that fires when the store triggers. And the store triggers when it updates.

File: frontend/src/components/sprintboard/task.jsx [hunk 6]
```diff
@@ -1,18 +1,78 @@
...
 });
 
-export default Task;
+Task.propTypes = {
+  connectDragSource: PropTypes.func.isRequired,
+  connectDropTarget: PropTypes.func.isRequired,
+  isDragging: PropTypes.bool.isRequired
+};
+
+export default DropTarget(ItemTypes.SPRINTTASK, taskTarget, collectForDrop)(DragSource(ItemTypes.SPRINTTASK, taskSource, collectForDrag)(Task));
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/task.jsx]: can you explain what this is doing?

File: frontend/src/components/sprintboard/board.jsx [hunk 7]
```diff
@@ -48,4 +53,5 @@ let SprintBoard = React.createClass({
 
 });
 
-export default SprintBoard;
+export {SprintBoard}; // exporting unwrapped component for testing purposes
+export default DragDropContext(HTML5Backend)(SprintBoard);
```
Comment (wezleytsai) [inline frontend/src/components/sprintboard/board.jsx:56]: does this actually need to be exported? or is it also for testing purposes

File: frontend/src/components/sprintboard/board.jsx (hunk 1 above)
Comment (wezleytsai) [inline frontend/src/components/sprintboard/board.jsx:25]: Yea, that's what i was thinking too.

Is it possible for the store to be triggering multiple kinds of data? For example, `trigger({a: a})` in one method, `trigger({b:b})` in another? And then have the component's listener check what got triggered? Or is that a bad pattern

Comment (azangru) [inline frontend/src/components/sprintboard/board.jsx:25]: I was very curious about that too. The only way I could think of was like you are describing: trigger an event, pass a payload, and then in a function inside the component, depending on what the payload is, execute other appropriate functions.

But it feels like the idea is that a store provides a component with all the information it needs to know in one single message.

http://stackoverflow.com/questions/29048164/can-refluxjs-stores-indicate-which-property-has-changed-when-they-call-trigger

**Pull request merged**